numpy/f2py/crackfortran.py

```diff
@@ -325,6 +330,28 @@ def is_free_format(file):
     return result
 
 
+def openhook(filename, mode):
+    """Ensures that filename is opened with correct encoding parameter.
+
+    This function uses chardet package, when available, for
+    determining the encoding of the file to be opened. When chardet is
+    not available, the function detects only UTF-8-SIG encoding,
+    otherwise, ASCII encoding is used.
+    """
+    bytes = min(32, os.path.getsize(filename))
+    with open(filename, 'rb') as f:
+        raw = f.read(bytes)
+    if raw.startswith(codecs.BOM_UTF8):
+        encoding = 'UTF-8-SIG'
+    else:
+        if chardet is not None:
+            encoding = chardet.detect(raw)['encoding']
+        else:
+            # hint: install chardet to ensure correct encoding handling
```

Comment:
Should this be documented somewhere else / added to the `environment.yml` with a unit test / something else? I'm not fully sure what the rules for adding dependencies (even optional ones) are.

Reply:
I would say that, say, 99.99..% of f2py inputs work fine with ascii encoding: this PR introduces non-ascii encoding support for the first time.

In other cases, the lack-of-chardet could be treated with a warning but in this case, it would mean that most f2py users will start seeing this warning while their inputs to f2py are actually just plain ascii files and the warning messages will be just noise to these users and forcing them to install chardet just to remove a warning message would be an overkill, imho. So, I don't recommend adding chardet to `environment.yml` as for the majority of f2py users it would be useless overhead.

To resolve this question, I added tests and implemented exception message that will suggest installing chardet when reading inputs using ascii fails.

Reply:
I have added support to UTF-16 and UTF-32 as well.

It turns out that detecting other encodings is not reliable even when using chardet, so its importance as a dependency to f2py has diminished considerably with the recent changes.